internal/controller/object/informers.go

```diff
@@ -0,0 +1,273 @@
+		for _, p := range ps {
+			if !p.Generic(ev) {
+				return
+			}
+		}
+		h.Generic(context.WithValue(ctx, keyProviderConfigName, providerConfig), ev, q)
+	}
+
+	go func() {
+		<-ctx.Done()
+		i.sink = nil
+	}()
+
+	return nil
+}
+
+// WatchResources starts informers for the given resource GVKs for the given
+// cluster (i.e. rest.Config & providerConfig).
+// The is wired into the Object reconciler, which will call this method on
+// every reconcile to make resourceInformers aware of the referenced or managed
+// resources of the given Object.
+//
+// Note that this complements garbageCollectResourceInformers which regularly
+// garbage collects resource informers that are no longer referenced by
+// any Object.
+func (i *resourceInformers) WatchResources(rc *rest.Config, providerConfig string, gvks ...schema.GroupVersionKind) {
+	if rc == nil {
+		rc = i.config
+	}
+
+	// start new informers
+	for _, gvk := range gvks {
+		i.lock.RLock()
+		_, found := i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}]
+		i.lock.RUnlock()
+		if found {
+			continue
+		}
+
+		log := i.log.WithValues("providerConfig", providerConfig, "gvk", gvk.String())
+
+		ca, err := cache.New(rc, cache.Options{})
+		if err != nil {
+			log.Debug("failed creating a cache", "error", err)
+			continue
+		}
+
+		// don't forget to call cancelFn in error cases to avoid leaks. In the
+		// happy case it's called from the go routine starting the cache below.
+		ctx, cancelFn := context.WithCancel(context.Background())
+
+		u := kunstructured.Unstructured{}
+		u.SetGroupVersionKind(gvk)
+		inf, err := ca.GetInformer(ctx, &u, cache.BlockUntilSynced(false)) // don't block. We wait in the go routine below.
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				obj := newObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				if old.GetResourceVersion() == obj.GetResourceVersion() {
+					return
+				}
+
+				ev := runtimeevent.GenericEvent{
+					Object: obj,
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			DeleteFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+		}); err != nil {
+			cancelFn()
+			log.Debug("failed adding event handler", "error", err)
+			continue
+		}
+
+		go func() {
+			defer cancelFn()
+
+			log.Info("Starting resource watch")
+			_ = ca.Start(ctx)
+		}()
+
+		i.lock.Lock()
+		i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}] = resourceCache{
```

Comment:
can't it be that we leak a running cache if there is a second racing call of `WatchResources`? Better double check here under the lock that the `resourceCache` is still not there, and if it is, stop the new one.